File: MOS/src/core/assembly_log.cpp
#include "mos/core/assembly_log.hpp"

#include "json.hpp"

#include <algorithm>
#include <chrono>
#include <fstream>
#include <stdexcept>

namespace mos {
namespace core {

namespace {

/// Comparator over node INDICES by the canonical key. Python sorts by
/// `nodes[i].key()`, a tuple; NodeRecord::operator< reproduces that ordering.
struct ByKey {
    const std::vector<NodeRecord>& nodes;
    bool operator()(int a, int b) const { return nodes[a] < nodes[b]; }
};

std::string join_support(const std::vector<int>& support) {
    std::string out;
    for (std::size_t i = 0; i < support.size(); ++i) {
        if (i) out += ',';
        out += std::to_string(support[i]);
    }
    return out;
}

double now_seconds() {
    const auto d = std::chrono::system_clock::now().time_since_epoch();
    return std::chrono::duration<double>(d).count();
}

}  // namespace

std::string canonical_signature(const std::vector<NodeRecord>& nodes,
                                const std::vector<std::pair<int, int>>& edges) {
    const int n = static_cast<int>(nodes.size());

    std::map<int, std::vector<int>> children;
    for (int i = 0; i < n; ++i) children[i] = {};
    std::vector<int> indeg(static_cast<std::size_t>(n), 0);

    for (const auto& e : edges) {
        const int p = e.first;
        const int c = e.second;
        if (!(0 <= p && p < n && 0 <= c && c < n)) {
            throw std::invalid_argument(
                "dependency (" + std::to_string(p) + "," + std::to_string(c) +
                ") out of range for " + std::to_string(n) + " nodes");
        }
        children[p].push_back(c);
        indeg[static_cast<std::size_t>(c)] += 1;
    }

    const ByKey by_key{nodes};

    // Python: sorted(i for i in range(n) if indeg[i] == 0). The generator yields
    // in increasing i, and sorted() is stable, so pushing in increasing i and
    // then stable_sort reproduces it exactly.
    std::vector<int> ready;
    for (int i = 0; i < n; ++i) {
        if (indeg[static_cast<std::size_t>(i)] == 0) ready.push_back(i);
    }
    std::stable_sort(ready.begin(), ready.end(), by_key);

    std::vector<int> order;
    order.reserve(static_cast<std::size_t>(n));
    while (!ready.empty()) {
        // Python pops from the FRONT, appends newly-ready children at the BACK,
        // then re-sorts the whole list with a stable sort. Ties therefore keep
        // already-present nodes ahead of newly-appended ones. Replicated here
        // because the resulting index map feeds the edge part of the signature,
        // so a different tie-break is a different signature.
        const int i = ready.front();
        ready.erase(ready.begin());
        order.push_back(i);

        for (const int c : children[i]) {
            if (--indeg[static_cast<std::size_t>(c)] == 0) ready.push_back(c);
        }
        std::stable_sort(ready.begin(), ready.end(), by_key);
    }

    if (static_cast<int>(order.size()) != n) {
        throw std::invalid_argument("composite DAG contains a cycle; refusing to sign it");
    }

    std::map<int, int> pos;
    for (std::size_t k = 0; k < order.size(); ++k) pos[order[k]] = static_cast<int>(k);

    std::string node_part;
    for (std::size_t k = 0; k < order.size(); ++k) {
        if (k) node_part += '|';
        const NodeRecord& nd = nodes[static_cast<std::size_t>(order[k])];
        node_part += nd.op_type;
        node_part += ':';
        node_part += join_support(nd.support);
        node_part += ':';
        node_part += (nd.read_only ? 'r' : 'w');
    }

    std::vector<std::string> edge_strings;
    edge_strings.reserve(edges.size());
    for (const auto& e : edges) {
        edge_strings.push_back(std::to_string(pos[e.first]) + ">" +
                               std::to_string(pos[e.second]));
    }
    // Python `sorted()` on str is lexicographic by code point; std::sort on
    // std::string is lexicographic by char. Identical over ASCII digits and '>'.
    std::sort(edge_strings.begin(), edge_strings.end());

    std::string edge_part;
    for (std::size_t k = 0; k < edge_strings.size(); ++k) {
        if (k) edge_part += ';';
        edge_part += edge_strings[k];
    }

    return node_part + "#" + edge_part;
}

AssemblyLog::AssemblyLog(std::string path) : path_(std::move(path)) {}

int AssemblyLog::record(std::vector<NodeRecord> nodes,
                        std::vector<std::pair<int, int>> edges,
                        std::vector<std::vector<int>> slices,
                        std::uint64_t tick,
                        std::optional<double> rho_before,
                        std::string note) {
    AssemblyEvent ev;
    ev.signature = canonical_signature(nodes, edges);
    ev.nodes = std::move(nodes);
    ev.edges = std::move(edges);
    ev.slices = std::move(slices);
    ev.tick = tick;
    ev.wall_time = now_seconds();
    ev.rho_before = rho_before;
    ev.note = std::move(note);

    const int handle = next_handle_++;
    open_.emplace(handle, std::move(ev));
    return handle;
}

void AssemblyLog::close(int handle,
                        std::optional<double> rho_after,
                        std::optional<std::string> verified,
                        const std::string& extra_note) {
    auto it = open_.find(handle);
    if (it == open_.end()) {
        throw std::invalid_argument("AssemblyLog::close on unknown handle " +
                                    std::to_string(handle));
    }
    AssemblyEvent ev = std::move(it->second);
    open_.erase(it);

    ev.rho_after = rho_after;
    ev.verified = std::move(verified);
    ev.closed = true;
    if (!extra_note.empty()) {
        ev.note = ev.note.empty() ? extra_note : ev.note + "; " + extra_note;
    }
    flush(ev);
}

void AssemblyLog::set_activation(int handle,
                                 std::vector<std::string> retrieved,
                                 std::vector<std::string> grown) {
    auto it = open_.find(handle);
    if (it == open_.end()) {
        throw std::invalid_argument(
            "AssemblyLog::set_activation on unknown handle " +
            std::to_string(handle));
    }
    it->second.retrieved = std::move(retrieved);
    it->second.grown = std::move(grown);
}

void AssemblyLog::abandon(int handle, const std::string& note) {
    auto it = open_.find(handle);
    if (it == open_.end()) {
        throw std::invalid_argument("AssemblyLog::abandon on unknown handle " +
                                    std::to_string(handle));
    }
    AssemblyEvent ev = std::move(it->second);
    open_.erase(it);

    ev.closed = false;  // explicit: the outcome was never learned
    ev.note = ev.note.empty() ? note : ev.note + "; " + note;
    flush(ev);
}

void AssemblyLog::flush(const AssemblyEvent& ev) {
    nlohmann::json j;
    j["signature"] = ev.signature;

    nlohmann::json nodes = nlohmann::json::array();
    for (const auto& n : ev.nodes) {
        nodes.push_back({{"op_type", n.op_type},
                         {"support", n.support},
                         {"read_only", n.read_only}});
    }
    j["nodes"] = std::move(nodes);

    nlohmann::json edges = nlohmann::json::array();
    for (const auto& e : ev.edges) {
        edges.push_back(nlohmann::json::array({e.first, e.second}));
    }
    j["edges"] = std::move(edges);

    j["slices"] = ev.slices;
    j["tick"] = ev.tick;
    j["wall_time"] = ev.wall_time;
    // Always emitted, even when empty. An absent key and an empty list would be
    // indistinguishable to the reader, and "this tick retrieved nothing" is a
    // real observation that a cover model needs -- it is a row of zeros, not a
    // missing row.
    j["retrieved"] = ev.retrieved;
    j["grown"] = ev.grown;

    // Absent measurements serialise as null, never as 0.0.
    j["rho_before"] = ev.rho_before ? nlohmann::json(*ev.rho_before) : nlohmann::json(nullptr);
    j["rho_after"] = ev.rho_after ? nlohmann::json(*ev.rho_after) : nlohmann::json(nullptr);
    j["verified"] = ev.verified ? nlohmann::json(*ev.verified) : nlohmann::json(nullptr);
    j["closed"] = ev.closed;
    j["note"] = ev.note;
    const auto d = ev.delta_rho();
    j["delta_rho"] = d ? nlohmann::json(*d) : nlohmann::json(nullptr);

    std::ofstream out(path_, std::ios::app);
    if (!out) {
        // A failed write is data loss the registry calls unrecoverable, so it is
        // surfaced rather than swallowed. It is not fatal to the tick.
        throw std::runtime_error("AssemblyLog: cannot append to " + path_);
    }
    out << j.dump() << "\n";
}

}  // namespace core
}  // namespace mos

File: MOS/python/assembly_log.py
"""
TATIANA — E7: record assembly AT COMPOSITION TIME, because it is unrecoverable later.

WHY THIS EXISTS, AND WHY ITS STATED PURPOSE CHANGED
---------------------------------------------------
E7 was specified (audit 3.3, book 48.3 item 5) as "record Ext^1 data when
composites are assembled". **That purpose is dead**: F10 is dissolved (logbook
5aa/5af), the K_0 / Jordan-Holder / Ext^1 memory schema is dropped, and there is
no longer any category in which to take an Ext group. Building an Ext recorder
now would be building for a theory we retired.

What survives, and is in fact the stronger reason, is the *instruction*:

    "record the raw assembly, decide what it means afterwards"
    "cheap now and impossible later"

because MOS's memory is **history-determined by construction** (the whole basis
for rejecting both FCA and canonical K_0 schemas). Two live mechanisms consume
assembly history and neither can reconstruct it after the fact:

  1. **delta-D promotion (Construction 3).** "A microneuron is a promoted operad
     sub-tree: when a composite DAG repeatedly resolves conflicts it is promoted
     from composite to first-class GENERATOR." That requires knowing which
     composites RECURRED and whether they actually resolved conflict. Both are
     composition-time facts.
  2. **Q(t) and Lambda(t).** Wiring-change and compression curves are time series;
     a gap in the series is permanent.

So this module records composition events, not extension classes. The
interpretation is deliberately left open, exactly as E7 instructed.

THE IDENTITY PROBLEM, AND WHY THE SIGNATURE IS A COMPLETE INVARIANT
--------------------------------------------------------------------
"Did this composite recur?" is only well-posed if composites have a well-defined
identity. Comparing DAGs is graph isomorphism in general, but ours are tiny and
TYPED, which collapses the difficulty: we serialise the dependency structure in a
canonical order obtained by topological sort with ties broken deterministically
on (op_type, sorted support).

That signature is a COMPLETE invariant for our purposes: two composites share a
signature exactly when they agree on every node's (type, support) AND on every
dependency edge between them -- which is precisely what it means to be the same
composite. Tie-breaking cannot conflate two genuinely different DAGs, because the
dependency edges are part of what is serialised, not just the node multiset.

APPEND-ONLY, AND WHY
--------------------
The file is opened in append mode and never rewritten. A consolidation pass may
later summarise it, but the raw record is the thing E7 says is unrecoverable, so
nothing in this module deletes or edits a past line. Same discipline as
e5_elicitations.jsonl, which already paid off once when E5's verdict thresholds
had to be replaced and the data could be re-scored at zero API cost (5v).
"""

from __future__ import annotations

import json
import os
import time
from dataclasses import dataclass, field, asdict
from typing import Dict, Iterable, List, Optional, Sequence, Tuple

DEFAULT_PATH = os.path.join(os.path.dirname(os.path.abspath(__file__)),
                            "assembly_events.jsonl")


@dataclass(frozen=True)
class NodeRecord:
    """One operator occurrence inside a composite."""
    op_type: str
    support: Tuple[int, ...]          # sorted; () means global
    read_only: bool

    def key(self) -> Tuple[str, Tuple[int, ...], bool]:
        return (self.op_type, self.support, self.read_only)


@dataclass
class AssemblyEvent:
    """One composition, recorded when it is assembled -- not when it is judged.

    `rho_before` is known at assembly time. `rho_after`, `verified` and
    `delta_rho` are filled in by `close()` once the composite has run; an event
    that never closes is recorded as incomplete rather than silently dropped,
    because "we assembled this and never learned the outcome" is itself data.
    """
    signature: str
    nodes: List[NodeRecord]
    edges: List[Tuple[int, int]]      # dependency (parent_idx, child_idx)
    slices: List[List[int]]           # foliation: node indices per slice
    tick: int
    wall_time: float
    rho_before: Optional[float] = None
    rho_after: Optional[float] = None
    verified: Optional[str] = None    # "VERIFIED" | "REFUTED" | "UNVERIFIABLE"
    closed: bool = False
    note: str = ""

    @property
    def delta_rho(self) -> Optional[float]:
        if self.rho_before is None or self.rho_after is None:
            return None
        return self.rho_after - self.rho_before

    def to_json(self) -> Dict:
        d = asdict(self)
        d["nodes"] = [{"op_type": n.op_type,
                       "support": list(n.support),
                       "read_only": n.read_only} for n in self.nodes]
        d["edges"] = [list(e) for e in self.edges]
        d["delta_rho"] = self.delta_rho
        return d


def canonical_signature(nodes: Sequence[NodeRecord],
                        edges: Sequence[Tuple[int, int]]) -> str:
    """A complete, order-independent invariant of the composite's structure.

    Topological sort with deterministic tie-breaking on (op_type, support,
    read_only), then serialise nodes AND the dependency edges rewritten in the
    canonical index space. Raises on a cycle -- an operad DAG with a cycle is a
    bug worth surfacing, not something to serialise around.
    """
    n = len(nodes)
    children: Dict[int, List[int]] = {i: [] for i in range(n)}
    indeg = [0] * n
    for (p, c) in edges:
        if not (0 <= p < n and 0 <= c < n):
            raise ValueError(f"dependency ({p},{c}) out of range for {n} nodes")
        children[p].append(c)
        indeg[c] += 1

    ready = sorted((i for i in range(n) if indeg[i] == 0),
                   key=lambda i: nodes[i].key())
    order: List[int] = []
    while ready:
        i = ready.pop(0)
        order.append(i)
        for c in children[i]:
            indeg[c] -= 1
            if indeg[c] == 0:
                ready.append(c)
        ready.sort(key=lambda j: nodes[j].key())

    if len(order) != n:
        raise ValueError("composite DAG contains a cycle; refusing to sign it")

    pos = {orig: k for k, orig in enumerate(order)}
    node_part = "|".join(
        f"{nodes[i].op_type}:{','.join(map(str, nodes[i].support))}"
        f":{'r' if nodes[i].read_only else 'w'}" for i in order)
    edge_part = ";".join(sorted(f"{pos[p]}>{pos[c]}" for (p, c) in edges))
    return f"{node_part}#{edge_part}"


class AssemblyLog:
    """Append-only recorder. Nothing here rewrites or deletes a past line."""

    def __init__(self, path: str = DEFAULT_PATH):
        self.path = path
        self._open: Dict[int, AssemblyEvent] = {}
        self._next_handle = 0

    def record(self,
               nodes: Sequence[NodeRecord],
               edges: Sequence[Tuple[int, int]],
               slices: Sequence[Sequence[int]],
               tick: int,
               rho_before: Optional[float] = None,
               note: str = "") -> int:
        """Record a composition AT ASSEMBLY TIME. Returns a handle for close()."""
        ev = AssemblyEvent(
            signature=canonical_signature(nodes, edges),
            nodes=list(nodes),
            edges=[tuple(e) for e in edges],
            slices=[list(s) for s in slices],
            tick=tick,
            wall_time=time.time(),
            rho_before=rho_before,
            note=note,
        )
        h = self._next_handle
        self._next_handle += 1
        self._open[h] = ev
        return h

    def close(self, handle: int,
              rho_after: Optional[float] = None,
              verified: Optional[str] = None) -> AssemblyEvent:
        """Fill in the outcome and flush the event to disk."""
        if handle not in self._open:
            raise KeyError(f"unknown assembly handle {handle}")
        ev = self._open.pop(handle)
        ev.rho_after = rho_after
        ev.verified = verified
        ev.closed = True
        self._flush(ev)
        return ev

    def abandon(self, handle: int, note: str = "abandoned") -> AssemblyEvent:
        """Flush an event whose outcome was never learned.

        Recorded as closed=False. "We assembled this and never found out" is a
        fact about the run, and dropping it would bias any later analysis toward
        composites that happened to finish.
        """
        if handle not in self._open:
            raise KeyError(f"unknown assembly handle {handle}")
        ev = self._open.pop(handle)
        ev.note = (ev.note + " | " + note).strip(" |")
        self._flush(ev)
        return ev

    def _flush(self, ev: AssemblyEvent) -> None:
        with open(self.path, "a", encoding="utf-8") as fh:
            fh.write(json.dumps(ev.to_json(), sort_keys=True) + "\n")

    # -- read side: what delta-D promotion will actually ask -------------------

    @staticmethod
    def load(path: str = DEFAULT_PATH) -> List[Dict]:
        if not os.path.exists(path):
            return []
        out: List[Dict] = []
        with open(path, "r", encoding="utf-8") as fh:
            for line in fh:
                line = line.strip()
                if line:
                    out.append(json.loads(line))
        return out

    @staticmethod
    def recurrence(events: Iterable[Dict]) -> Dict[str, int]:
        """signature -> how many times that composite has been assembled."""
        counts: Dict[str, int] = {}
        for e in events:
            counts[e["signature"]] = counts.get(e["signature"], 0) + 1
        return counts

    @staticmethod
    def promotion_evidence(events: Iterable[Dict]) -> Dict[str, Dict]:
        """Per signature: the evidence Construction 3's promotion gate needs.

        Deliberately returns EVIDENCE, not a decision. Per selection.py's rule,
        delta_rho alone must never promote (coherent-but-wrong), so the verify
        tallies are reported separately and the caller applies the gate.
        """
        acc: Dict[str, Dict] = {}
        for e in events:
            s = acc.setdefault(e["signature"], {
                "assembled": 0, "closed": 0, "verified": 0, "refuted": 0,
                "unverifiable": 0, "delta_rhos": []})
            s["assembled"] += 1
            if e.get("closed"):
                s["closed"] += 1
            v = e.get("verified")
            if v == "VERIFIED":
                s["verified"] += 1
            elif v == "REFUTED":
                s["refuted"] += 1
            elif v == "UNVERIFIABLE":
                s["unverifiable"] += 1
            if e.get("delta_rho") is not None:
                s["delta_rhos"].append(e["delta_rho"])
        return acc


# ---------------------------------------------------------------------------
# Self-tests. Run: python assembly_log.py
# ---------------------------------------------------------------------------

def _main() -> None:
    import tempfile

    ok = 0

    def check(name: str, cond: bool) -> None:
        nonlocal ok
        assert cond, f"FAILED: {name}"
        ok += 1
        print(f"  [ok] {name}")

    print("=== E7: assembly recording ===")

    A = NodeRecord("SearchOp", (), True)
    B = NodeRecord("ReasonOp", (1,), False)
    C = NodeRecord("VerifyOp", (2,), False)

    # 1. The signature is order-independent -- the same composite described with
    #    its nodes listed differently must sign identically, or "did this recur?"
    #    is meaningless.
    s1 = canonical_signature([A, B, C], [(0, 1), (1, 2)])
    s2 = canonical_signature([C, B, A], [(2, 1), (1, 0)])
    check("signature is independent of node listing order", s1 == s2)

    # 2. ... but it is NOT blind to structure: same nodes, different dependency
    #    edges must sign differently.
    s3 = canonical_signature([A, B, C], [(0, 1), (0, 2)])
    check("different dependency structure signs differently", s1 != s3)

    # 3. ... nor to support or type.
    s4 = canonical_signature([A, NodeRecord("ReasonOp", (9,), False), C],
                             [(0, 1), (1, 2)])
    check("different support signs differently", s1 != s4)
    s5 = canonical_signature([A, NodeRecord("ComputeOp", (1,), False), C],
                             [(0, 1), (1, 2)])
    check("different op type signs differently", s1 != s5)

    # 4. A cycle is a bug, not something to serialise around.
    try:
        canonical_signature([A, B], [(0, 1), (1, 0)])
        raise AssertionError("FAILED: cycle was not rejected")
    except ValueError:
        ok += 1
        print("  [ok] a cyclic 'DAG' is refused rather than signed")

    with tempfile.TemporaryDirectory() as td:
        path = os.path.join(td, "assembly.jsonl")
        log = AssemblyLog(path)

        # 5. Record at assembly time, close at outcome time.
        h = log.record([A, B, C], [(0, 1), (1, 2)], [[0], [1], [2]],
                       tick=1, rho_before=0.40)
        check("nothing is written before the outcome is known",
              not os.path.exists(path) or len(AssemblyLog.load(path)) == 0)
        ev = log.close(h, rho_after=0.12, verified="VERIFIED")
        check("delta_rho is computed, not stored twice",
              abs(ev.delta_rho - (0.12 - 0.40)) < 1e-12)
        check("one line on disk after close", len(AssemblyLog.load(path)) == 1)

        # 6. An abandoned composite is still recorded. Dropping it would bias
        #    later analysis toward composites that happened to finish.
        h2 = log.record([A, B], [(0, 1)], [[0, 1]], tick=2, rho_before=0.5)
        log.abandon(h2, "engine crashed")
        rows = AssemblyLog.load(path)
        check("an abandoned composite is recorded as closed=False",
              len(rows) == 2 and rows[1]["closed"] is False)
        check("and its delta_rho is None, not 0.0",
              rows[1]["delta_rho"] is None)

        # 7. Append-only: an existing file is added to, never rewritten.
        log2 = AssemblyLog(path)
        h3 = log2.record([A, B, C], [(0, 1), (1, 2)], [[0], [1], [2]],
                         tick=3, rho_before=0.33)
        log2.close(h3, rho_after=0.30, verified="UNVERIFIABLE")
        rows = AssemblyLog.load(path)
        check("a new AssemblyLog appends rather than truncating", len(rows) == 3)

        # 8. Recurrence -- the delta-D question. The tick-1 and tick-3 composites
        #    are structurally identical and must be counted together.
        counts = AssemblyLog.recurrence(rows)
        check("the recurring composite is counted twice", counts[s1] == 2)
        check("the one-off composite is counted once",
              sum(v for k, v in counts.items() if k != s1) == 1)

        # 9. Promotion EVIDENCE, not a promotion decision.
        ev_map = AssemblyLog.promotion_evidence(rows)
        e1 = ev_map[s1]
        check("verify outcomes are tallied separately from delta_rho",
              e1["verified"] == 1 and e1["unverifiable"] == 1
              and e1["refuted"] == 0)
        check("both delta_rhos are retained for the caller to gate on",
              len(e1["delta_rhos"]) == 2)

        # 10. Unknown handles are errors, not silent no-ops.
        try:
            log.close(999)
            raise AssertionError("FAILED: unknown handle accepted")
        except KeyError:
            ok += 1
            print("  [ok] closing an unknown handle raises")

    print(f"\nAll {ok} assembly-log tests passed.")


if __name__ == "__main__":
    _main()
